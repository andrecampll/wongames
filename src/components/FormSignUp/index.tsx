/* eslint-disable no-unused-expressions */
import React, { useState } from 'react';
import Link from 'next/link';
import { signIn } from 'next-auth/client';
import { AccountCircle, Email, Lock } from '@styled-icons/material-outlined';
import { useMutation } from '@apollo/client';
import { MUTATION_REGISTER } from '../../graphql/mutations/register';
import { UsersPermissionsRegisterInput } from '../../graphql/generated/globalTypes';

import Button from '../Button';
import TextField from '../TextField';

import { FormWrapper, FormLink } from '../Form';

const FormSignUp = () => {
  const [values, setValues] = useState<UsersPermissionsRegisterInput>({
    username: '',
    email: '',
    password: '',
  });

  const [register, { error }] = useMutation(MUTATION_REGISTER, {
    onError: err => err,
    onCompleted: () => {
      !error &&
        signIn('credentials', {
          email: values.email,
          password: values.password,
          callbackUrl: '/',
        });
    },
  });

  const handleInput = async (field: string, value: string) => {
    setValues(s => ({ ...s, [field]: value }));
  };

  const handleSubmit = async (event: React.FormEvent) => {
    event.preventDefault();

    register({
      variables: {
        input: {
          username: values.username,
          email: values.email,
          password: values.password,
        },
      },
    });
  };

  return (
    <FormWrapper>
      <form onSubmit={handleSubmit}>
        <TextField
          name="name"
          placeholder="Name"
          type="name"
          onInputChange={v => handleInput('username', v)}
          icon={<AccountCircle />}
        />
        <TextField
          name="email"
          placeholder="Email"
          type="email"
          onInputChange={v => handleInput('email', v)}
          icon={<Email />}
        />
        <TextField
          name="password"
          placeholder="Password"
          type="password"
          onInputChange={v => handleInput('password', v)}
          icon={<Lock />}
        />
        <TextField
          name="confirm-password"
          placeholder="Confirm password"
          type="password"
          onInputChange={v => handleInput('password-confirm', v)}
          icon={<Lock />}
        />

        <Button type="submit" size="large" fullWidth>
          Sign up now
        </Button>

        <FormLink>
          Already have an account?{' '}
          <Link href="/sign-in">
            <a>Sign in</a>
          </Link>
        </FormLink>
      </form>
    </FormWrapper>
  );
};

export default FormSignUp;
