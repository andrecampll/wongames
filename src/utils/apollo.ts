import { useMemo } from 'react';
import { setContext } from '@apollo/client/link/context';
import {
  ApolloClient,
  HttpLink,
  NormalizedCache,
  NormalizedCacheObject,
} from '@apollo/client';
import { Session } from 'next-auth';
import apolloCache from './apolloCache';

let apolloClient: ApolloClient<NormalizedCacheObject | null>;

function createApolloClient(session?: Session | null) {
  const httpLink = new HttpLink({
    uri: `${process.env.NEXT_PUBLIC_API_URL}/graphql`,
  });

  const authLink = setContext((_, { headers, session: clientSession }) => {
    const jwt = session?.jwt || clientSession?.jwt || '';
    const authorization = jwt ? `Bearer ${jwt}` : '';

    return { headers: { ...headers, authorization } };
  });

  return new ApolloClient({
    ssrMode: typeof window === 'undefined',
    link: authLink.concat(httpLink),
    cache: apolloCache,
  });
}

export function initializeApollo(
  initialState = null,
  session?: Session | null,
) {
  const apolloClientGlobal = apolloClient ?? createApolloClient(session);

  if (initialState) {
    apolloClientGlobal.cache.restore(
      initialState as NormalizedCacheObject & NormalizedCache,
    );
  }

  if (typeof window === 'undefined') return apolloClientGlobal;

  apolloClient = apolloClient ?? apolloClientGlobal;

  return apolloClient;
}

export function useApollo(initialState = null, session?: Session) {
  const store = useMemo(() => initializeApollo(initialState, session), [
    initialState,
    session,
  ]);

  return store;
}
