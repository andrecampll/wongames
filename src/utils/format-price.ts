export function formatPrice(price: number | bigint): string {
  return new Intl.NumberFormat('en', {
    style: 'currency',
    currency: 'USD',
  }).format(price);
}
