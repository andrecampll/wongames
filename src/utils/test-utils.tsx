import { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';

import { ThemeProvider } from 'styled-components';

import {
  WishlistContext,
  WishlistContextDefaultValues,
  WishlistContextDTO,
} from '../hooks/wishlist/useWishlist';

import theme from '../styles/theme';
import {
  CartContext,
  CartContextDTO,
  CartContextDefaultValues,
} from '../hooks/cart/useCart';

type CustomRenderProps = {
  cartProviderProps?: CartContextDTO;
  wishlistProviderProps?: WishlistContextDTO;
} & Omit<RenderOptions, 'queries'>;

const customRender = (
  ui: ReactElement,
  {
    cartProviderProps = CartContextDefaultValues,
    wishlistProviderProps = WishlistContextDefaultValues,
    ...renderOptions
  }: CustomRenderProps = {},
) =>
  render(
    <ThemeProvider theme={theme}>
      <CartContext.Provider value={cartProviderProps}>
        <WishlistContext.Provider value={wishlistProviderProps}>
          {ui}
        </WishlistContext.Provider>
      </CartContext.Provider>
    </ThemeProvider>,
    renderOptions,
  );

export * from '@testing-library/react';
export { customRender as render };
