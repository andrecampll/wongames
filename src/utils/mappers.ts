import { QueryGames_games } from '../graphql/generated/QueryGames';
import {
  QueryHome_banners,
  QueryHome_sections_freeGames_highlight,
} from '../graphql/generated/QueryHome';
import { QueryWishlist_wishlists_games } from '../graphql/generated/QueryWishlist';

import { formatPrice } from './format-price';

export const bannerMapper = (banners: QueryHome_banners[]) => {
  return banners.map(({ image, title, subtitle, button, ribbon }) => ({
    image: `http://localhost:1337${image?.url}`,
    title,
    subtitle,
    buttonLabel: button?.label,
    buttonLink: button?.link,
    ...(ribbon && {
      ribbon: ribbon.text,
      ribbonColor: ribbon.color,
      ribbonSize: ribbon.size,
    }),
  }));
};

export const gamesMapper = (games: QueryGames_games[] | null | undefined) => {
  return games
    ? games.map(({ id, name, slug, developers, cover, price }) => ({
        id,
        title: name,
        slug,
        developer: developers[0].name,
        image: cover?.url
          ? `http://localhost:1337${cover?.url}`
          : 'https://i.stack.imgur.com/y9DpT.jpg',
        price,
      }))
    : [];
};

export const highlightMapper = (
  highlight: QueryHome_sections_freeGames_highlight | null | undefined,
) => {
  return highlight
    ? {
        title: highlight.title,
        subtitle: highlight.subtitle,
        backgroundImage: `http://localhost:1337${highlight.background.url}`,
        floatImage: `http://localhost:1337${highlight.floatImage.url}`,
        buttonLabel: highlight.buttonLabel,
        buttonLink: highlight.buttonLink,
        alignment: highlight.alignment,
      }
    : {};
};

export const cartMapper = (
  games:
    | QueryGames_games[]
    | QueryWishlist_wishlists_games[]
    | null
    | undefined,
) => {
  return games
    ? games.map(game => ({
        id: game.id,
        img: `http://localhost:1337${game.cover?.url}`,
        title: game.name,
        price: formatPrice(game.price),
      }))
    : [];
};
