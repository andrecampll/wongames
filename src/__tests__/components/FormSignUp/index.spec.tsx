import { MockedProvider } from '@apollo/client/testing';
import { screen } from '@testing-library/react';
import FormSignUp from '../../../components/FormSignUp';
import { render } from '../../../utils/test-utils';

describe('<FormSignUp />', () => {
  it('should render the form', () => {
    const { container } = render(
      <MockedProvider>
        <FormSignUp />
      </MockedProvider>,
    );

    expect(screen.getByPlaceholderText(/name/i)).toBeInTheDocument();
    expect(screen.getByPlaceholderText(/email/i)).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Password')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Confirm password')).toBeInTheDocument();
    expect(
      screen.getByRole('button', { name: /sign up now/i }),
    ).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render text and link to sign in', () => {
    render(
      <MockedProvider>
        <FormSignUp />
      </MockedProvider>,
    );
    expect(screen.getByRole('link', { name: /sign in/i })).toBeInTheDocument();
    expect(screen.getByText(/already have an account\?/i)).toBeInTheDocument();
  });
});
