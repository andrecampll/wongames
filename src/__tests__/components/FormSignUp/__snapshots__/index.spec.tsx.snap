// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormSignUp /> should render the form 1`] = `
<div
  class="sc-iBPRYJ hlxcom"
>
  <form>
    <div
      class="sc-jSgupP eplsmf"
    >
      <div
        class="sc-bdfBwQ OPvhY"
      >
        <div
          class="sc-hKgILt Ddcon"
        >
          <svg
            aria-hidden="true"
            class="StyledIconBase-ea9ulj-0 cKCFNq"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zM7.07 18.28c.43-.9 3.05-1.78 4.93-1.78s4.51.88 4.93 1.78C15.57 19.36 13.86 20 12 20s-3.57-.64-4.93-1.72zm11.29-1.45c-1.43-1.74-4.9-2.33-6.36-2.33s-4.93.59-6.36 2.33A7.95 7.95 0 014 12c0-4.41 3.59-8 8-8s8 3.59 8 8c0 1.82-.62 3.49-1.64 4.83zM12 6c-1.94 0-3.5 1.56-3.5 3.5S10.06 13 12 13s3.5-1.56 3.5-3.5S13.94 6 12 6zm0 5c-.83 0-1.5-.67-1.5-1.5S11.17 8 12 8s1.5.67 1.5 1.5S12.83 11 12 11z"
            />
          </svg>
        </div>
        <input
          class="sc-gsTCUz gcCRQr"
          name="name"
          placeholder="Name"
          type="name"
          value=""
        />
      </div>
    </div>
    <div
      class="sc-jSgupP eplsmf"
    >
      <div
        class="sc-bdfBwQ OPvhY"
      >
        <div
          class="sc-hKgILt Ddcon"
        >
          <svg
            aria-hidden="true"
            class="StyledIconBase-ea9ulj-0 cKCFNq"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M22 6c0-1.1-.9-2-2-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6zm-2 0l-8 5-8-5h16zm0 12H4V8l8 5 8-5v10z"
            />
          </svg>
        </div>
        <input
          class="sc-gsTCUz gcCRQr"
          name="email"
          placeholder="Email"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="sc-jSgupP eplsmf"
    >
      <div
        class="sc-bdfBwQ OPvhY"
      >
        <div
          class="sc-hKgILt Ddcon"
        >
          <svg
            aria-hidden="true"
            class="StyledIconBase-ea9ulj-0 cKCFNq"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
            >
              <path
                d="M0 0h24v24H0V0z"
              />
              <path
                d="M0 0h24v24H0V0z"
                opacity="0.87"
              />
            </g>
            <path
              d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
            />
          </svg>
        </div>
        <input
          class="sc-gsTCUz gcCRQr"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      class="sc-jSgupP eplsmf"
    >
      <div
        class="sc-bdfBwQ OPvhY"
      >
        <div
          class="sc-hKgILt Ddcon"
        >
          <svg
            aria-hidden="true"
            class="StyledIconBase-ea9ulj-0 cKCFNq"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
            >
              <path
                d="M0 0h24v24H0V0z"
              />
              <path
                d="M0 0h24v24H0V0z"
                opacity="0.87"
              />
            </g>
            <path
              d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
            />
          </svg>
        </div>
        <input
          class="sc-gsTCUz gcCRQr"
          name="confirm_password"
          placeholder="Confirm password"
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      class="sc-gKsewC laRVME"
      type="submit"
    >
      <span>
        <span>
          Sign up now
        </span>
      </span>
    </button>
    <div
      class="sc-kEjbxe jIJKxq"
    >
      Already have an account?
       
      <a
        href="/sign-in"
      >
        Sign in
      </a>
    </div>
  </form>
</div>
`;
